File: fetch.py
import requests
import logging as logger
from pathlib import Path
import argparse
from datetime import datetime, timedelta

BASE_URL = 'http://markets.cboe.com/us/equities/market_statistics/historical_market_volume'


def fetch(yyyy):
    fname = f'market_history_{yyyy}.csv'
    url = f'{BASE_URL}/{fname}-dl'

    fpath = Path(__file__).resolve().parent / "data"
    fpath.mkdir(parents=True, exist_ok=True)
    ffile = fpath / fname

    req = requests.get(url)
    with ffile.open('w') as f:
        f.write(req.text)
    logger.info(f'Saving data to {ffile}')


def get_years(date1, date2):
    year1 = date1.year
    year2 = date2.year + 1
    for y in range(year1, year2):
        yield y


if __name__ == '__main__':
    logger.basicConfig(
        level=logger.INFO,
        format='%(asctime)s %(levelname)-5.5s [%(name)s] %(message)s'
    )

    parser = argparse.ArgumentParser(description='Download daily volume data')
    parser.add_argument('--startDate',
                        dest='startDate',
                        action='store',
                        default=None,
                        help='Start date of the data period. Format: yyyymmdd'
                        )
    parser.add_argument('--endDate',
                        dest='endDate',
                        action='store',
                        default=None,
                        help='End date of the data period. Format: yyyymmdd'
                        )
    args = parser.parse_args()
    if args.startDate is None:
        start_date = datetime.today() - timedelta(days=6)
    else:
        start_date = datetime.strptime(args.startDate, '%Y%m%d')

    if args.endDate is None:
        end_date = datetime.today()
    else:
        end_date = datetime.strptime(args.endDate, '%Y%m%d')

    logger.info(f'startDate={start_date}, endDate={end_date}')

    years = get_years(start_date, end_date)
    for year in years:
        fetch(year)

File: fetch.js
const path = require('path');
const axios = require('axios');

const fs = require('fs-extra');
const writeFile = require('util').promisify(fs.writeFile);
const { createLogger, format, transports } = require('winston');
const argv = require('minimist')(process.argv.slice(2));

const logFormat = format.printf(
  (info) => `${info.timestamp} ${info.level} [${info.label}]: ${info.message}`
);
const logger = createLogger({
  format: format.combine(
    format.label({ label: 'DailyVolume' }),
    format.timestamp(),
    logFormat
  ),
  transports: [new transports.Console()],
});

async function download(baseURL, outfile) {
  logger.info(`Started downloading from ${baseURL}`);
  const api = axios.create({
    baseURL,
  });

  try {
    const res = await api.get();
    const fpath = path.dirname(outfile);
    fs.ensureDirSync(fpath);
    if (res.data) {
      await writeFile(outfile, res.data);
      logger.info(`Finished writing to ${outfile}`);
    } else {
      throw res.data;
    }
  } catch (error) {
    logger.error(`Failed in downloading from url: ${baseURL}`);
    logger.error(`Found error: ${error}`);
  }
}

function getFilepath() {
  return path.join(__dirname, 'data');
}

function getFilename(yyyy) {
  return `market_history_${yyyy}.csv`;
}

function downloadYear(yyyy) {
  const baseUrl = 'http://markets.cboe.com/us/equities/market_statistics';
  const url = `${baseUrl}/historical_market_volume`;
  const fname = getFilename(yyyy);
  const link = `${url}/${fname}-dl`;
  const fpath = getFilepath();
  const outfile = path.join(fpath, fname);
  download(link, outfile);
}

function getYears(startDate, endDate) {
  console.log(typeof startDate);
  console.log(typeof endDate);
  const startYear = +startDate.toString().slice(0, 4);
  const endYear = +endDate.toString().slice(0, 4);
  const years = [];
  for (let year = startYear; year <= endYear; year += 1) {
    years.push(year);
  }
  return years;
}

const d = new Date();
const endDate = argv.endDate === undefined ? d.getFullYear().toString() : argv.endDate;
d.setDate(d.getDate() - 6);
const startDate = argv.startDate === undefined ? d.getFullYear().toString() : argv.startDate;
logger.info(`startDate=${startDate}, endDate=${endDate}`)

const years = getYears(startDate, endDate);
console.log(years);
years.forEach(function (year) {
  console.log(year);
  downloadYear(year);
});
